py3status/modules/playerctl.py

```diff
@@ -0,0 +1,268 @@
+from gi.repository import GLib, Playerctl
+
+
+class Py3status:
+    """ """
+
+    # available configuration parameters
+    button_next = None
+    button_pause = None
+    button_play = None
+    button_play_pause = 1
+    button_previous = None
+    button_stop = 3
+    cache_timeout = 5
+    format = "{format_player}"
+    format_player = (
+        r"[\?if=is_playing > ][\?if=is_paused \|\| ]"
+        r"[\?if=is_stopped .. ][[{artist}][\?soft  - ][{title}]]"
+    )
+    format_separator = " "
+    players = []
+    sleep_timeout = 20
+
+    def post_config_hook(self):
+        if not self.py3.check_commands("playerctl"):
+            raise Exception("command 'playerctl' not installed")
+
+        self.color_paused = self.py3.COLOR_PAUSED or self.py3.COLOR_DEGRADED
+        self.color_playing = self.py3.COLOR_PLAYING or self.py3.COLOR_GOOD
+        self.color_stopped = self.py3.COLOR_STOPPED or self.py3.COLOR_BAD
+
+        self._init_manager()
+
+    def _init_manager(self):
+        self.manager = Playerctl.PlayerManager()
+        self.manager.connect("name-appeared", self._on_name_appeared)
+        self.manager.connect("name-vanished", self._on_name_vanished)
+
+        # Initialize all currently active players
+        for player_name in self.manager.props.player_names:
+            self._init_player(player_name)
+
+        self._start_loop()
+
+    def _init_player(self, player_name):
+        if len(self.players) > 0 and player_name.name not in self.players:
+            return
+
+        player = Playerctl.Player.new_from_name(player_name)
+
+        # Set up all event listeners
+        player.connect("playback-status", self._status_changed, self.manager)
+        player.connect("metadata", self._on_metadata, self.manager)
+
+        self.manager.manage_player(player)
+
+    def _start_loop(self):
+        loop = GLib.MainLoop()
+
+        # The loop is blocking so it needs to be run a separate thread
+        thread = Thread(target=loop.run)
+        thread.daemon = True
+        thread.start()
+
+    def _on_name_appeared(self, manager, player_name):
+        self._init_player(player_name)
+
+    def _on_name_vanished(self, manager, name):
+        self.py3.update()
+
+    def _on_metadata(self, player, metadata, manager):
+        self.py3.update()
+
+    def _status_changed(self, player, status, manager):
+        self.py3.update()
+
+    def _get_player_position(self, player):
+        try:
+            # Playerctl doesn't support getting positions for all players
+            microseconds = player.get_position()
+            position = self._microseconds_to_time(microseconds)
+        except GLib.GError:
+            position = None
+
+        return position
+
+    @staticmethod
+    def _microseconds_to_time(microseconds):
+        seconds = microseconds // 1_000_000
+        m, s = divmod(seconds, 60)
+        h, m = divmod(m, 60)
+        time = f"{h}:{m:02d}:{s:02d}"
+        return time.lstrip("0").lstrip(":")
```

Comment:
I agree. It is currently used in `cmus`, `battery_level` (a tiny bit different but it probably should change to this format), and this module. Currently, `spotify` and `mpris` use Python's `datetime.timedelta` (which this function could actually use because it does the same thing. It just needs to use `lstrip` twice as well) to convert it into a valid time so they could change to this py3 function.